src/frontend/Semantic_check.ml

```diff
@@ -669,9 +669,10 @@ and semantic_check_funapp ~is_cond_dist id es cf emeta =
     ok
       { e with
         expr=
-          ( match e.expr with
-          | FunApp (fun_kind, id, ues) -> CondDistApp (fun_kind, id, ues)
-          | _ -> raise_s [%sexp ("This should never happen!" : string)] ) })
+          ( match (e.expr, is_cond_dist) with
+          | FunApp (fun_kind, id, ues), true -> CondDistApp (fun_kind, id, ues)
```

Comment:
Why does this step happen? Shouldn't we just keep the AST the same structurally? Isn't your check supposed to check for that?

Reply:
Because the FunApp and ConDistApp use lots of the same checking functions. Those are currently set up to return a FunApp. The alternative is to thread the is_cond_dist argument through all of them and have them return ConDistApps when appropriate to start with, rather than having to correct for that at the end. Would that be preferable?

Reply:
I've gone with that alternative now. It also seems a bit cleaner to me.

Reply:
Ahh, okay. If this is where the error was happening before, it seems like a good refactoring to make it cleaner now. Will take another look, thanks!